### PR TITLE
Project Structure Update

### PR DESCRIPTION
**Pull request by KnightTec**
This PR updates the project structure for better clarity. This includes:
- seperation of C++ code into the `Source` directory
- shader are left in the top level directory as they are technically not part of the application source but resources used by it
- consistent UpperCamelCase naming of all directories and C++ files (except for vsg files)
- renamed the pictures directory to Screenshots as this is a more descriptive name

The `Source` directory has the following structure:
- `Buffers`: contains all render buffer definitions
- `External`: same as before; vsg code
- `IO`
- `RenderModules`: all classes that create rendering commands; all of those have a similar structure and might be candidates for a common interface
- `Scene`: everything related to the scene (graph), its modification and scene descriptor creation

If accepted, should be merged after #17.

### CODE REVIEW

**Comment by Lachei** [comment]
Did not compile out of the box as well as errors occurred when trying to find the shaders (Executable directory was in the Source folder, as the add_executable call in cmake was done in this directory).
Pulled the add_executable call out to the top level cmake as well as added the source folder to the include directories. Changed includes in the files to use <> with relative paths to the Source folder.
@KnightTec pls check and verify compilation and validity of program.

**Review by Lachei — APPROVED** [review]

**Comment by KnightTec** [comment]
Had to set cmake to use C++ 17 to compile on windows.
@Lachei please verify that it still compiles on linux before I merge.

**Comment by KnightTec** [comment]
@Lachei do you still know how you got git subtree to work properly? It seems to be broken with the new directory stucture. I tried this solution ([https://stackoverflow.com/questions/21702996/git-subtree-add-change-prefix-preserving-local-commits](https://stackoverflow.com/questions/21702996/git-subtree-add-change-prefix-preserving-local-commits)) but this does not keep the changes we made to vsg and vsgXchange.

**Comment by Lachei** [comment]
Hmm. Just tried this solution https://stackoverflow.com/questions/46055656/git-subtree-move-subtree-to-a-different-directory-and-pull-it but also deletes all local changes.
Gonna see if i find something else

**Comment by Lachei** [comment]
Ok, so the error is that i added it with squashed history as it seems.
However do not yet know how to fix...

**Comment by KnightTec** [comment]
When Source/External is moved back to external subtree works again.
Maybe there is a way to then move external to Source/External including the git subtree prefix.

**Comment by Lachei** [comment]
Interesting. And the subtree split into extra branch and then adding that extra branch via git subtree add with the correct prefix didnt work?

> do you still know how you got git subtree to work properly? 

When i added the subtree via subtree add i did encounter no problems...

**Comment by KnightTec** [comment]
> Interesting. And the subtree split into extra branch and then adding that extra branch via git subtree add with the correct prefix didnt work?

After moving Source/External back to external I ran the following commands:
```
$ git subtree split --prefix=external/vsg HEAD
2edf6c17d9b43c5f877aadccd4178c9af0669df3
$ git branch split_vsg 2edf6c17d9b43c5f877aadccd4178c9af0669df3
$ git rm -r external/vsg
$ git commit -m "Removed vsg"
$ git subtree add --prefix=Source/External/vsg . 2edf6c17d9b43c5f877aadccd4178c9af0669df3
$ git subtree pull --prefix Source/External/vsg git@github.com:vsg-dev/VulkanSceneGraph.git master --squash
```
The last `subtree pull` still overwrites the local history.

**Comment by KnightTec** [comment]
One thing that I thought might be a problem is that, after renaming from `external` to `Source/External`, the history is not tracked correctly anymore (according to git log --follow Source/External). So I did a history preserving rename using this [git-vm-with-history](https://gist.github.com/emiller/6769886#file-git-mv-with-history) command.
Afterwards the renamed files had the correct history.
Then I ran
`git subtree split --rejoin --prefix=Source/External/vsg HEAD`
and the history was still correct.
But running
`git subtree pull --prefix=Source/External/vsg git@github.com:vsg-dev/VulkanSceneGraph.git master --squash`
in the end still overwrote all local changes.

I am really out of ideas here. @Lachei can we just merge the PR for now so I can work on the build scripts and CI pipeline? And fix the subtree problem in another PR.

**Comment by Lachei** [comment]
We could of course merge this PR for now, however if there should be no solution to our problem and we have to cherry pick our changes all further changes done after the PR might also be overwritten and would require remerging.
Thus i would recommend to instead simply make a new branch from this branch and do further implementation for the CI pipeline on that branch. This way at least it is clear that this branch is not yet bugfree.
Also, as we have other forks of this repo, we definitely should make sure that such bugs that might overwrite other work are not on the main branch.

**Comment by Lachei** [comment]
Still did not find a good solution. Currently have tried out a merge via `git pull -s subtree -Xsubtree=Source/External/vsg git@github.com:vsg-dev/VulkanSceneGraph.git master --allow-unrelated-histories` which seems to then allow `git subtree pull --prefix Source/External/vsg git@github.com:vsg-dev/VulkanSceneGraph.git master` to properly check for updates.
This however adds ALL commits from the vulkan scene graph repo which is also undesirable (We are talking about 3200 commits).

**Comment by Lachei** [comment]
@KnightTec thinking about the restructuring i am wondering:

Is the here proposed structure really a good one?

When i think about it i would it find better, to simply put the buffers and io folder in the top level folder instead of building up a Source folder. This would not require us to move the external folder at all and would be still very well usable. Maybe another folder will be added for pipelines, but that would be about it.
Extracting a src folder as in this PR is done also would require an extra header folder (at least that is what my intuition says). Which makes quickly adding new stuff often way more complicated down the line. Also i am not convinced that camel case for the folder is the correct way to go. I think small case as is already done for external in the master branch is more familiar.This of course requires renaming of the denoiser and taa folder, but that would be easy enough.

So basically my idea would be to go with the current master and just add 2 or three folders for accumulation buffer, pipelines and only leave the VulkanPBRT.cpp file in the top level directory. This would of course also completley erase the issues with subtree...

**Comment by Lachei** [comment]
Also i would like to add that for the other projects that fork from this would i think be better off if we did not alter the external folder...

**Comment by KnightTec** [comment]
> Still did not find a good solution. Currently have tried out a merge via `git pull -s subtree -Xsubtree=Source/External/vsg git@github.com:vsg-dev/VulkanSceneGraph.git master --allow-unrelated-histories` which seems to then allow `git subtree pull --prefix Source/External/vsg git@github.com:vsg-dev/VulkanSceneGraph.git master` to properly check for updates. This however adds ALL commits from the vulkan scene graph repo which is also undesirable (We are talking about 3200 commits). Resolving the merge conflicts at least for vsg is manageable and needed about half an hour. For vsgXchange i suspect that i will need a bit longer. @KnightTec should we use this solution then as a fix?

If we squash merge the `directory_structure_00` branch into `master` then our `master` commit history would stay clean. I think then we could use this fix.

**Comment by Lachei** [comment]
> If we squash merge the directory_structure_00 branch into master then our master commit history would stay clean. I think then we could use this fix.

Yes that would at least keep our history cleaner. However i do not know if this then screws up subtree again as it can again not match the commits of the external repo to the local commit history... I definitely want to prevent a second merge by hand!
Also, the problem of forks having to port their code to the new location still stands.
However i guess you defenitely want to have a separate src folder, according to missing comments on the other ideas? If so what is the reason to use camel case in the first place for folders?

**Comment by Lachei** [comment]
> Yes that would at least keep our history cleaner. However i do not know if this then screws up subtree again as it can again not match the commits of the external repo to the local commit history...

It does indeed produce an even more significant error: even after rejoining the subtree one can not pull via subtree pull. This would mean that we might have to merge by hand every time we want to update....

**Comment by KnightTec** [comment]
> @KnightTec thinking about the restructuring i am wondering:
> 
> Is the here proposed structure really a good one?
> 
> When i think about it i would it find better, to simply put the buffers and io folder in the top level folder instead of building up a Source folder. This would not require us to move the external folder at all and would be still very well usable. Maybe another folder will be added for pipelines, but that would be about it. Extracting a src folder as in this PR is done also would require an extra header folder (at least that is what my intuition says). Which makes quickly adding new stuff often way more complicated down the line. Also i am not convinced that camel case for the folder is the correct way to go. I think small case as is already done for external in the master branch is more familiar.This of course requires renaming of the denoiser and taa folder, but that would be easy enough.
> 
> So basically my idea would be to go with the current master and just add 2 or three folders for accumulation buffer, pipelines and only leave the VulkanPBRT.cpp file in the top level directory. This would of course also completley erase the issues with subtree...
> 
> Also i would like to ask @chrismile what he thinks about this.

The top level directory will also contain multiple build scripts and other files related to those. Having a Source folder offers a clean seperation between our source code and everything else. It would also prevent our top level directory from getting too cluttered if we add more source related directories. 
I added the `Scene` directory so we can implement our own scene management system and put the code in there instead of waiting on a better vsg implementation of the scene graph with actual geometry instead of raster commands. 

`Going slightly off-topic in the following`
Even if vsg would provide a better implementation tomorrow I think we should still go with our scene manager.
The way vsg forces literally everything into a graph structure (and with it the needlessly complicated visitor pattern) just does not make a lot of sense.
Even the scene graph itself is not really fit as a primary storage structure of a scene. It should just represent the parent-child relationships of our objects. If we want to do frustum culling we would nedd another stucture anyway such as an Quadtree/Octree.
A better way would be to store our objects indepedently from these hierarchical structures and have any tree/graph structure refer to the objects using an object ID.
`Off-topic end`

Regarding the casing of folder names:
Before this PR the repo contained a mix of both lowercamelcase and UpperCamelCase so I went with the latter to keep the folder naming consistent with the .cpp and .h files.

Regarding the seperate include directory:
Seperating headers and source files is useful when the project is used as a library. Then package managers can easily download only the headers and the compiled library file without the `.cpp` source (thats how vcpkg does it). But since our project is not used as a library there is no real need for an include folder. 
The [UE4 repo](https://github.com/EpicGames/UnrealEngine) for example has an Engine/Source folder that contains both header and implementation files.

**Comment by KnightTec** [comment]
> It does indeed produce an even more significant error: even after rejoining the subtree one can not pull via subtree pull. This would mean that we might have to merge by hand every time we want to update....

I think before we put way more time into fixing the subtree problem we could just leave the `external` folder at the top level for now.

**Comment by Lachei** [comment]
> I think before we put way more time into fixing the subtree problem we could just leave the external folder at the top level for now.

I just talked with @chrismile and he also sugested this as he likes his external libraries at the top level :).
I also think that this would be the best approach then, having the source folder with all our source files and the external folder at the same level as the source folder.
There was no real reason behind the camel case naming of some folders. I personally like lowercase folders better.

I think that the refactor of the assimp import will still take time so we will eventually do the implementation our selves..

@KnightTec could you then maybe change the code structure to this proposal?

**Comment by KnightTec** [comment]
> 
> I just talked with @chrismile and he also sugested this as he likes his external libraries at the top level :). I also think that this would be the best approach then, having the source folder with all our source files and the external folder at the same level as the source folder. There was no real reason behind the camel case naming of some folders. I personally like lowercase folders better.
> 
> I think that the refactor of the assimp import will still take time so we will eventually do the implementation our selves..
> 
> @KnightTec could you then maybe change the code structure to this proposal?

Will do.

**Comment by KnightTec** [comment]
> Will do.

Done. Can be merged after one final compile check on linux.

**Comment by Lachei** [comment]
I would say we should use lowercase for all folders to stay consistent with the external folder. What do you think @KnightTec ?

**Comment by Lachei** [comment]
I will do the renaming if you approve.

**Comment by KnightTec** [comment]
> I would say we should use lowercase for all folders to stay consistent with the external folder. What do you think @KnightTec ?

Sounds good.

**Comment by Lachei** [comment]
Ok done, will merge now.
